ocaml/xapi/xapi_xenops.ml

```diff
@@ -2275,13 +2278,29 @@ let update_vm ~__context id =
               try
                 Option.iter
                   (fun (_, state) ->
-                    debug "xenopsd event: Updating VM %s last_start_time <- %s"
-                      id
-                      (Date.to_string (Date.of_float state.Vm.last_start_time)) ;
                     let metrics = Db.VM.get_metrics ~__context ~self in
                     let start_time = Date.of_float state.Vm.last_start_time in
-                    Db.VM_metrics.set_start_time ~__context ~self:metrics
-                      ~value:start_time ;
+                    if
+                      start_time
+                      <> Db.VM_metrics.get_start_time ~__context ~self:metrics
+                    then (
+                      debug
+                        "xenopsd event: Updating VM %s last_start_time <- %s" id
+                        (Date.to_string (Date.of_float state.Vm.last_start_time)) ;
+                      Db.VM_metrics.set_start_time ~__context ~self:metrics
+                        ~value:start_time ;
+                      if
+                        (* VM reboot *)
+                        power_state = `Running
+                        && power_state_before_update = `Running
```

Comment:
Why do we need this condition? I believe we found that the start_time is only updated on start or reboot.

With this condition, we potentially miss the case where, during a reboot the power_state is temporarily set to Halted. This depends on when the state is checked.

Reply:
From test, I see the `start_time` is updated on VM start, VM restart, VM resume. And updated to "19700101T00:00:00Z" on VM suspend and VM shutdown.

Reply:
From my test, this always matches a reboot and it also passed xenrt tests. OK, I see, this may potentially miss some VM reboot.
Then how about changing to the last implementation, update on "VM start", "VM shutdown", "VM restart" by filtering "VM suspend" and "VM resume":
`
                      if
                        power_state <> `Suspended (* not VM suspend *)
                        && power_state_before_update
                           <> `Suspended (* not VM resume *)
                      then
                        (* clear RestartVM guidance on VM start, VM shutdown,
                           VM restart *)
                        Xapi_vm_helpers.remove_pending_guidance ~__context ~self
                          ~value:`restart_vm
`

Reply:
The transition to 1970 for suspended VMs is unfortunate... Then I'd do
```
power_state = `Running
&& (power_state_before_update <> `Suspended)
```

Reply:
OK, I see, can you please help to confirm we will clear the VM guidance both when VM is "down"(VM shutdown, VM suspend) and when VM is "up" (VM start, VM reboot):

```
power_state = `Running
&& (power_state_before_update <> `Suspended)
```
would match "VM start", "VM restart", and we can clear ``restart_device_model`  and ``restart_vm` for the 2 "VM up" scenarios.

And with the code in `force_state_reset_keep_current_operations`:
```
  if state = `Suspended then
    remove_pending_guidance ~__context ~self ~value:`restart_device_model ;
  if state = `Halted then (	  if state = `Halted then (
    remove_pending_guidance ~__context ~self ~value:`restart_device_model ;
    remove_pending_guidance ~__context ~self ~value:`restart_vm
    ) ;
```
which is for "VM down": clears ``restart_device_model` on "VM suspend" and "VM shutdown", and clears ``restart_vm` on "VM shutdown".

Thank you!

Reply:
I don't mind that... call it "defence in depth" :)